spec/gemstash/s3_storage_spec.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+require "yaml"
+require "aws-sdk-s3"
+
+RSpec.describe Gemstash::S3 do
+  before(:all) do
+    env = Gemstash::Env.new
+    Gemstash::Env.current = env
+    env.config = Gemstash::Configuration.new
+    env.config[:s3_path => "TEMP_S3_TEST_FOLDER/gemstash/s3_storage"]
+    @storage = Gemstash::S3.for("private").for("gems")
+    @folder = "gemstash/s3_storage"
+  end
+  after(:all) do
+    VCR.use_cassette("batch delete objects") do
+      @storage.delete_with_prefix
+    end
+  end
+
+  context "intialize storage component" do
+    it("builds with a valid folder") { expect(Gemstash::S3.new(@folder)).not_to be_nil }
+    it "stores metadata about Gemstash and the storage engine version" do
+      expect(Gemstash::S3.metadata[:storage_version]).to eq(Gemstash::S3::STORAGE_VERSION)
+      expect(Gemstash::S3.metadata[:gemstash_version]).to eq(Gemstash::VERSION)
+    end
+
+    it "prevents using storage engine if the storage version is too new" do
+      metadata = {
+        storage_version: 999_999,
+        gemstash_version: Gemstash::VERSION
+      }
+      File.write(Gemstash::Env.current.base_file("metadata.yml"), metadata.to_yaml)
+      expect { Gemstash::S3.new(@folder) }.
+        to raise_error(Gemstash::S3::VersionTooNew)
+    end
+  end
+
+  context "with a valid storage" do
+    let(:gem_contents) { read_gem("example", "0.1.0") }
+    before(:context) do
+    end
+    it "can create a child storage from itself" do
+      child_storage = @storage.for("gems")
+      expect(child_storage).to be_truthy
+      expect(child_storage).to be_instance_of Gemstash::S3
+      expect(child_storage.folder).to eq(File.join(@storage.folder, "gems"))
+    end
+
+    it "returns a non existing resource when requested", :vcr do
+      resource = @storage.resource("an_id")
+      expect(resource).not_to be_nil
+      expect(resource).not_to exist
```

Comment:
I think we can just test "not to exist"

Reply:
What do you mean by this? ^ Isn't already test not_to exist?

Reply:
Yeah, if we are testing for that do you think we still need to test for nil too?

Reply:
[](url)
```suggestion
      expect(resource). be__of Gemstash::S3::Resource
```

I think we still need to test for nil because even if the resource doesn't exist (in this context means it has no content), the resource object shouldn't be nil. I think I should add this line?